File: 93.ModularInverseAlgo.py
# A modular inverse of a modulo m is a number x such that (a * x) % m == 1. 
# It exists iff gcd(a, m) == 1 (i.e., a and m are coprime). 
# For prime m we can compute it quickly with exponentiation using 
# Fermat’s little theorem.

# Explanation (clear, step-by-step)

# The modular inverse x satisfies a * x ≡ 1 (mod m).

# If gcd(a, m) != 1, no inverse exists because a shares a common factor with m and 
# you cannot get 1 as a linear combination.

# Two common algorithms:

# Extended Euclidean Algorithm (EEA) — finds integers x, y such that 
# a*x + m*y = gcd(a,m). If gcd == 1, x (mod m) is the inverse. Works for any m.

# Fermat’s Little Theorem — if m is prime, inverse = a^(m-2) mod m (fast with modular exponentiation).

# EEA is general and safe for non-prime modulus; Fermat's method is faster when modulus is prime.

def extended_gcd(a, b):
    if b == 0:
        return (a, 1, 0)
    g, x1, y1 = extended_gcd(b, a % b)
    x = y1
    y = x1 - (a // b) * y1
    return (g, x, y)

def modinv_eea(a, m):
    g, x, _ = extended_gcd(a, m)
    if g != 1:
        return None
    return x % m

def modinv_fermat(a, m):
    return pow(a, m - 2, m)

# Demo examples and printing output
def demo():
    examples = [
        (3, 11),   # coprime, prime modulus: inverse exists (should be 4)
        (10, 17),  # coprime, prime modulus: inverse exists (should be 12)
        (6, 9)     # gcd != 1, inverse does not exist
    ]

    for a, m in examples:
        print(f"Example: a = {a}, m = {m}")
        inv1 = modinv_eea(a, m)
        if inv1 is None:
            print("  EEA: No modular inverse (gcd != 1)")
        else:
            print(f"  EEA: modular inverse = {inv1}  (check: {(a * inv1) % m} )")
        # use Fermat only when modulus is prime — here we demonstrate when it's safe
        if is_probably_prime(m):
            inv2 = modinv_fermat(a, m)
            print(f"  Fermat: modular inverse = {inv2}  (check: {(a * inv2) % m} )")
        else:
            print("  Fermat: skipped (modulus not prime)")
        print()

# Simple (deterministic for small m) primality check for demo purposes
def is_probably_prime(n):
    if n < 2:
        return False
    if n % 2 == 0:
        return n == 2
    r = 3
    while r * r <= n:
        if n % r == 0:
            return False
        r += 2
    return True

# Run demo
if __name__ == "__main__":
    demo()

File: 93.ModularInverseAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

public class ModularInverse {

    // Extended Euclidean Algorithm:
    // Returns gcd(a, b) and finds coefficients x, y such that a*x + b*y = gcd(a, b)
    static long[] extendedGCD(long a, long b) {
        // base case: when b == 0, gcd = a and coefficients are (1, 0)
        if (b == 0) return new long[]{a, 1, 0};

        // recursive call
        long[] res = extendedGCD(b, a % b);

        // compute x, y using results from recursive call
        long g = res[0];       // gcd
        long x1 = res[1];      // coefficient for b
        long y1 = res[2];      // coefficient for a % b

        long x = y1;
        long y = x1 - (a / b) * y1;

        return new long[]{g, x, y};
    }

    // Modular Inverse using Extended Euclidean Algorithm (works for all modulus)
    static Long modInverseEEA(long a, long m) {
        long[] res = extendedGCD(a, m);
        long g = res[0];
        long x = res[1];

        // inverse exists only if gcd(a, m) == 1
        if (g != 1) return null;

        // x might be negative — normalize it under modulo m
        return (x % m + m) % m;
    }

    // Modular Inverse using Fermat's Little Theorem (only valid when m is prime)
    static long modInverseFermat(long a, long m) {
        return modPow(a, m - 2, m); // a^(m-2) mod m
    }

    // Fast Modular Exponentiation
    static long modPow(long base, long exp, long mod) {
        long result = 1;
        base %= mod;

        while (exp > 0) {
            if ((exp & 1) == 1)
                result = (result * base) % mod;
            base = (base * base) % mod;
            exp >>= 1;
        }
        return result;
    }

    // Simple primality check for demonstration
    static boolean isPrime(long n) {
        if (n < 2) return false;
        if (n % 2 == 0) return n == 2;
        for (long i = 3; i * i <= n; i += 2)
            if (n % i == 0)
                return false;
        return true;
    }

    // Demo function
    public static void main(String[] args) throws Exception {
        BufferedWriter out = new BufferedWriter(new OutputStreamWriter(System.out));

        long[][] examples = {
                {3, 11},  // Expected inverse = 4
                {10, 17}, // Expected inverse = 12
                {6, 9}    // gcd != 1, no inverse
        };

        for (long[] e : examples) {
            long a = e[0], m = e[1];
            out.write("Example: a = " + a + ", m = " + m + "\n");

            Long inv1 = modInverseEEA(a, m);
            if (inv1 == null) {
                out.write("  EEA: No modular inverse (gcd != 1)\n");
            } else {
                out.write("  EEA: modular inverse = " + inv1 + "  (check: " + (a * inv1 % m) + " )\n");
            }

            if (isPrime(m)) {
                long inv2 = modInverseFermat(a, m);
                out.write("  Fermat: modular inverse = " + inv2 + "  (check: " + (a * inv2 % m) + " )\n");
            } else {
                out.write("  Fermat: skipped (modulus not prime)\n");
            }
            out.write("\n");
        }

        out.flush();
    }
}
